
    Clean columns names.
    Extract and convert numeric values from string values.
    Extract string data.
    Work with missing values.


In [2]:
import pandas as pd
laptops = pd.read_csv('/home/kamlesh/AnacondaProjects/Dataquest/Data/laptops.csv', encoding ="Latin-1" )

In [3]:
laptops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
Manufacturer                1303 non-null object
Model Name                  1303 non-null object
Category                    1303 non-null object
Screen Size                 1303 non-null object
Screen                      1303 non-null object
CPU                         1303 non-null object
RAM                         1303 non-null object
 Storage                    1303 non-null object
GPU                         1303 non-null object
Operating System            1303 non-null object
Operating System Version    1133 non-null object
Weight                      1303 non-null object
Price (Euros)               1303 non-null object
dtypes: object(13)
memory usage: 132.4+ KB


In [4]:
laptops.head(2)

Manufacturer   Model Name   Category Screen Size  \
0        Apple  MacBook Pro  Ultrabook       13.3"   
1        Apple  Macbook Air  Ultrabook       13.3"   

                               Screen                   CPU  RAM  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz  8GB   
1                            1440x900  Intel Core i5 1.8GHz  8GB   

               Storage                           GPU Operating System  \
0            128GB SSD  Intel Iris Plus Graphics 640            macOS   
1  128GB Flash Storage        Intel HD Graphics 6000            macOS   

  Operating System Version  Weight Price (Euros)  
0                      NaN  1.37kg       1339,69  
1                      NaN  1.34kg        898,94

In [5]:
print(laptops.columns)

Index(['Manufacturer', 'Model Name', 'Category', 'Screen Size', 'Screen',
       'CPU', 'RAM', ' Storage', 'GPU', 'Operating System',
       'Operating System Version', 'Weight', 'Price (Euros)'],
      dtype='object')


instructions

    Define a function, which accepts a string argument, and:
        Removes any whitespace from the start and end of the string.
        Replaces the substring Operating System with the abbreviation os.
        Replaces all spaces with underscores.
        Removes parentheses from the string.
        Makes the entire string lowercase.
        Returns the modified string.
    Use list comprehension to apply the function to each item in the DataFrame.column attribute for the laptops dataframe, assigning the result back to the the DataFrame.columns attribute.



    Defined a function, which:
        Used the str.strip() method to remove whitespace from the start and end of the string.
        Used the str.replace() method to remove parentheses from the string.
        Used the str.lower() method to make the string lowercase.
        Returns the modified string.
    Used list comprehension to apply the function to each item in the index object and assign it back to the DataFrame.columns attribute.
    Printed the new values for the DataFrame.columns attribute.



In [6]:
def clean_col(col):
    col = col.strip()
    col = col.replace("Operating System","os" )
    col = col.replace("(","")
    col = col.replace(")","")
    col = col.replace(" ","_" ) #spaces with underscores
  
    col = col.lower()
    return col

laptops.columns = [ clean_col(c)  for c in laptops.columns ]
laptops.columns

Index(['manufacturer', 'model_name', 'category', 'screen_size', 'screen',
       'cpu', 'ram', 'storage', 'gpu', 'os', 'os_version', 'weight',
       'price_euros'],
      dtype='object')

In [9]:
#just have a look at the columns
print(laptops.iloc[:5,2:5])

    category screen_size                              screen
0  Ultrabook       13.3"  IPS Panel Retina Display 2560x1600
1  Ultrabook       13.3"                            1440x900
2   Notebook       15.6"                   Full HD 1920x1080
3  Ultrabook       15.4"  IPS Panel Retina Display 2880x1800
4  Ultrabook       13.3"  IPS Panel Retina Display 2560x1600


### Observe the data type ... from the above outpt

Of these three columns, we have three different types of text data:

    category - This is purely text data, there are no numeric values.
    screen_size - This is numeric data that is being stored as text data because of the " character.
    screen - This is a combination of pure text data with numeric data.

Thinking about the screen_size column specifically, while the values are stored as text data, we can't sort them numerically. For instance, if we wanted to select laptops with screens 15" or larger we're unable to do so. To be able to answer these sorts of questions we need to convert the string values to numeric values.

In [13]:
print('Printing data type of Laptops screens colmn which is = ' + str(laptops["screen_size"].dtype))
print("\n")
print("Now printing datas/columns unique values")
print(laptops["screen_size"].unique())

Printing data type of Laptops screens colmn which is = object


Now printing datas/columns unique values
['13.3"' '15.6"' '15.4"' '14.0"' '12.0"' '11.6"' '17.3"' '10.1"' '13.5"'
 '12.5"' '13.0"' '18.4"' '13.9"' '12.3"' '17.0"' '15.0"' '14.1"' '11.3"']


Observe the pattern a single string ie is ' " ' need to be removed to have out columns as integer values

In [68]:
laptops["screen_size"] = laptops["screen_size"].str.replace('"','')


    Now we've removed the non-digit characters, we can convert (or cast) the column to a numeric dtype. To do this, we use the Series.astype() method. We can use either int or float as the parameter for the method to convert the column to the respective type:

In [69]:
laptops["screen_size"] = laptops["screen_size"].astype(float)

In [70]:
print(laptops["screen_size"].dtype)

float64


In [71]:

print(laptops["screen_size"].unique())

[13.3 15.6 15.4 14.  12.  11.6 17.3 10.1 13.5 12.5 13.  18.4 13.9 12.3
 17.  15.  14.1 11.3]


In [72]:
laptops.rename({"screen_size": "screen_size_inches"}, axis=1, inplace=True)

In [73]:
print(laptops.dtypes)

manufacturer           object
model_name             object
category               object
screen_size_inches    float64
screen                 object
cpu                    object
ram                    object
storage                object
gpu                    object
os                     object
os_version             object
weight                 object
price_euros            object
dtype: object


### Converting String Columns to Numeric
    1 Use the Series.replace() method to remove the substring GB from the ram column.
    2 Use the Series.astype() method to change the ram column to an integer dtype.
    3 Because the GB characters contained useful information about the units (gigabytes) of the laptop's
        ram, use the DataFrame.rename() method to rename the column from ram to ram_gb.
    4 Use the DataFrame.dtypes attribute to get a list of the column names and types from the laptop
    dataframe, and assign the result to dtypes.
    After you have run your code, use the variable inspector to view the dtypes variable to see the 
     results of your code.


In [84]:
#laptops.columns.tolist
laptops.head(2)

#below lines print the unique value in the ram column
print(laptops['ram'].unique())

#now printing the data type of the ram columns although it seems to be object but
print(laptops['ram'].dtype)

['8GB' '16GB' '4GB' '2GB' '12GB' '6GB' '32GB' '24GB' '64GB']
object


In [85]:
laptops["ram"] = laptops["ram"].str.replace('GB','').astype(int)

laptops.rename({"ram": "ram_gb"}, axis=1, inplace=True)

dtypes = laptops.dtypes
dtypes

manufacturer           object
model_name             object
category               object
screen_size_inches    float64
screen                 object
cpu                    object
ram_gb                  int64
storage                object
gpu                    object
os                     object
os_version             object
weight                 object
price_euros            object
dtype: object


    Clean the weight column by:
        Removing all non-digit characters.
        Casting the column to the appropriate numeric type
        Renaming the column to weight_kg.
    Clean the price_euros column by:
        Replacing the decimal commas with decimal points.
        Casting the column to the appropriate numeric type.
    Use the Series.describe() method to generate some descriptive statistics for each column:
        Assign the results for the weight_kg column to weight_describe.
        Assign the results for the price_euros column to price_describe


In [86]:
#below lines print the unique value in the ram column
print(laptops['weight'].unique())

#now printing the data type of the ram columns although it seems to be object but
print(laptops['weight'].dtype)

['1.37kg' '1.34kg' '1.86kg' '1.83kg' '2.1kg' '2.04kg' '1.3kg' '1.6kg'
 '2.2kg' '0.92kg' '1.22kg' '0.98kg' '2.5kg' '1.62kg' '1.91kg' '2.3kg'
 '1.35kg' '1.88kg' '1.89kg' '1.65kg' '2.71kg' '1.2kg' '1.44kg' '2.8kg'
 '2kg' '2.65kg' '2.77kg' '3.2kg' '0.69kg' '1.49kg' '2.4kg' '2.13kg'
 '2.43kg' '1.7kg' '1.4kg' '1.8kg' '1.9kg' '3kg' '1.252kg' '2.7kg' '2.02kg'
 '1.63kg' '1.96kg' '1.21kg' '2.45kg' '1.25kg' '1.5kg' '2.62kg' '1.38kg'
 '1.58kg' '1.85kg' '1.23kg' '1.26kg' '2.16kg' '2.36kg' '2.05kg' '1.32kg'
 '1.75kg' '0.97kg' '2.9kg' '2.56kg' '1.48kg' '1.74kg' '1.1kg' '1.56kg'
 '2.03kg' '1.05kg' '4.4kg' '1.90kg' '1.29kg' '2.0kg' '1.95kg' '2.06kg'
 '1.12kg' '1.42kg' '3.49kg' '3.35kg' '2.23kg' '4.42kg' '2.69kg' '2.37kg'
 '4.7kg' '3.6kg' '2.08kg' '4.3kg' '1.68kg' '1.41kg' '4.14kg' '2.18kg'
 '2.24kg' '2.67kg' '2.14kg' '1.36kg' '2.25kg' '2.15kg' '2.19kg' '2.54kg'
 '3.42kg' '1.28kg' '2.33kg' '1.45kg' '2.79kg' '1.84kg' '2.6kg' '2.26kg'
 '3.25kg' '1.59kg' '1.13kg' '1.78kg' '1.10kg' '1.15kg' '1.27kg' '1.43kg

In [88]:
laptops["weight"] = (laptops["weight"].str.replace("kg","").astype(float))

#error expected as some r kgs and kg

ValueError: could not convert string to float: '4s'

In [89]:
laptops["weight"] = laptops["weight"].str.replace('kgs','')
laptops["weight"] = laptops["weight"].str.replace('kg','').astype(float)

laptops["price_euros"] = laptops["price_euros"].str.replace(',','.').astype( float )

#renaming
laptops.rename({"weight":"weight_kg"},axis = 1,inplace = True)

weight_describe = laptops["weight_kg"].describe()

price_describe = laptops["price_euros"].describe()

In [90]:
laptops.head(2)

manufacturer   model_name   category  screen_size_inches  \
0        Apple  MacBook Pro  Ultrabook                13.3   
1        Apple  Macbook Air  Ultrabook                13.3   

                               screen                   cpu  ram_gb  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz       8   
1                            1440x900  Intel Core i5 1.8GHz       8   

               storage                           gpu     os os_version  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS        NaN   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS        NaN   

   weight_kg  price_euros  
0       1.37      1339.69  
1       1.34       898.94

Extract the manufacturer name from the cpu column, and assign it to a new column cpu_manufacturer.

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html

### Check the splitting strategy

In [14]:
laptops["cpu"].head()

0          Intel Core i5 2.3GHz
1          Intel Core i5 1.8GHz
2    Intel Core i5 7200U 2.5GHz
3          Intel Core i7 2.7GHz
4          Intel Core i5 3.1GHz
Name: cpu, dtype: object

In [15]:
laptops["cpu"].head().str.split() # seperated on whitespace

0           [Intel, Core, i5, 2.3GHz]
1           [Intel, Core, i5, 1.8GHz]
2    [Intel, Core, i5, 7200U, 2.5GHz]
3           [Intel, Core, i7, 2.7GHz]
4           [Intel, Core, i5, 3.1GHz]
Name: cpu, dtype: object

In [19]:
laptops["cpu"].head().str.split(n=1) # 1 split only

0          [Intel, Core i5 2.3GHz]
1          [Intel, Core i5 1.8GHz]
2    [Intel, Core i5 7200U 2.5GHz]
3          [Intel, Core i7 2.7GHz]
4          [Intel, Core i5 3.1GHz]
Name: cpu, dtype: object

In [23]:
laptops["cpu"].head().str.split(n=1,expand=True)

0                     1
0  Intel        Core i5 2.3GHz
1  Intel        Core i5 1.8GHz
2  Intel  Core i5 7200U 2.5GHz
3  Intel        Core i7 2.7GHz
4  Intel        Core i5 3.1GHz

In [26]:
# extrating the first column after all the operations
laptops["cpu_manufacture"] = laptops["cpu"].str.split(n=1,
                                                      expand=True).iloc[:,0]
                                                      

In [28]:
laptops["gpu_manufacturer"] = (laptops["gpu"]
                                    .str.split(n=1,expand=True)
                                    .iloc[:,0]
                               )

In [29]:
print(laptops["screen"].unique().shape)
print(laptops["screen"].unique()[:10])

(40,)
['IPS Panel Retina Display 2560x1600' '1440x900' 'Full HD 1920x1080'
 'IPS Panel Retina Display 2880x1800' '1366x768'
 'IPS Panel Full HD 1920x1080' 'IPS Panel Retina Display 2304x1440'
 'IPS Panel Full HD / Touchscreen 1920x1080'
 'Full HD / Touchscreen 1920x1080' 'Touchscreen / Quad HD+ 3200x1800']


There are 40 unique values, so we'll just look at the first 10. Some of the values have just the resolution (eg '1440x900'), while others have information on the type of screen and then end with the resolution. While the resolution is mostly numeric, we're going to treat the resolution as as a substring that we want to extract. Let's start by seeing what we get when we use Series.str.split():

In [94]:
print(laptops.loc[:9, "screen"].str.split(expand=True))

          0      1          2        3          4
0       IPS  Panel     Retina  Display  2560x1600
1  1440x900   None       None     None       None
2      Full     HD  1920x1080     None       None
3       IPS  Panel     Retina  Display  2880x1800
4       IPS  Panel     Retina  Display  2560x1600
5  1366x768   None       None     None       None
6       IPS  Panel     Retina  Display  2880x1800
7  1440x900   None       None     None       None
8      Full     HD  1920x1080     None       None
9       IPS  Panel       Full       HD  1920x1080


Five of the rows have the resolution in column 4, two have it in column 2 and the rest have it in column 1. Because the value we want to extract is from the end of the string and not the beginning of the string, the n=1 parameter is not going to help us. Luckily the Series.str.rsplit() method allows us to split from the end of the string instead of the front.

Let's look at a quick example to see the different between Series.str.rsplit() and Series.str.split().

In [30]:
print( laptops.loc[:9,"screen"].str.rsplit(n=1, expand=True))

                          0          1
0  IPS Panel Retina Display  2560x1600
1                  1440x900       None
2                   Full HD  1920x1080
3  IPS Panel Retina Display  2880x1800
4  IPS Panel Retina Display  2560x1600
5                  1366x768       None
6  IPS Panel Retina Display  2880x1800
7                  1440x900       None
8                   Full HD  1920x1080
9         IPS Panel Full HD  1920x1080


In [36]:
screen_res = laptops["screen"].str.rsplit(n=1, expand=True)
#
screen_res.head(5)

0          1
0  IPS Panel Retina Display  2560x1600
1                  1440x900       None
2                   Full HD  1920x1080
3  IPS Panel Retina Display  2880x1800
4  IPS Panel Retina Display  2560x1600

In [37]:

screen_res.columns = ["A", "B"]
screen_res.head()

A          B
0  IPS Panel Retina Display  2560x1600
1                  1440x900       None
2                   Full HD  1920x1080
3  IPS Panel Retina Display  2880x1800
4  IPS Panel Retina Display  2560x1600

In [38]:
# for rows where the value of column "B" is null, fill in the
# value found in column "A" for that row
screen_res.loc[screen_res["B"].isnull(), "B"] = screen_res["A"]
print(screen_res.iloc[:10])

                          A          B
0  IPS Panel Retina Display  2560x1600
1                  1440x900   1440x900
2                   Full HD  1920x1080
3  IPS Panel Retina Display  2880x1800
4  IPS Panel Retina Display  2560x1600
5                  1366x768   1366x768
6  IPS Panel Retina Display  2880x1800
7                  1440x900   1440x900
8                   Full HD  1920x1080
9         IPS Panel Full HD  1920x1080


In [39]:
#We've now extracted the resolution data, and we can then assign this to a new column:
laptops["screen_resolution"] = screen_res["B"]
print(laptops["screen_resolution"].unique().shape)
print(laptops["screen_resolution"].unique())

(15,)
['2560x1600' '1440x900' '1920x1080' '2880x1800' '1366x768' '2304x1440'
 '3200x1800' '1920x1200' '2256x1504' '3840x2160' '2160x1440' '2560x1440'
 '1600x900' '2736x1824' '2400x1600']


### Now get the processor speed like 2.3 only this

In [40]:

laptops["cpu"].unique().shape
laptops["cpu"].unique()[:6]

array(['Intel Core i5 2.3GHz', 'Intel Core i5 1.8GHz',
       'Intel Core i5 7200U 2.5GHz', 'Intel Core i7 2.7GHz',
       'Intel Core i5 3.1GHz', 'AMD A9-Series 9420 3GHz'], dtype=object)

In [41]:
pro_speed = laptops["cpu"].str.rsplit(n=1, expand=True)

In [42]:
pro_speed.columns = ["A","B"]

In [43]:
pro_speed["B"] = pro_speed['B'].str.replace('GHz','').astype(float)


In [104]:
laptops["cpu_speed_ghz"] = pro_speed["B"]

In [44]:
#ANother method                                    
laptops["cpu_speed_ghz"] = (laptops["cpu"]
                            .str.replace("GHz","")
                            .str.rsplit(n=1,expand=True)
                            .iloc[:,1]
                            .astype(float)
                            )


If your data has been scraped from a webpage, or if there was manual data entry involved at some point, you may end up with inconsistent values. Let's look at an example from our os column:

In [45]:
print(laptops["os"].value_counts())

Windows      1125
No OS          66
Linux          62
Chrome OS      27
macOS          13
Mac OS          8
Android         2
Name: os, dtype: int64




We can see that there are two variations on how the Apple operating system macOS exists in our dataset: Mac OS and macOS. One way we could fix this is by using a boolean comparison and assignment, but instead we'll learn a new way: the Series.map() method. The Series.map() method is ideal when we want to change multiple values in a column. Even though that's not the case here, we'll use it as an opportunity to learn how the method works.

The most common way to use Series.map() is with a dictionary. The keys of our dictionary are the original values in our series, and the corresponding values are what they're updated to. Let's look at a simple example using a series of misspelled fruit:


In [49]:
mapping_dict = {
    'Android': 'Android',
    'Chrome OS': 'Chrome OS',
    'Linux': 'Linux',
    'Mac OS': 'macOS',
    'No OS': 'No OS',
    'Windows': 'Windows',
    'macOS': 'macOS'
}

laptops["os"] = laptops["os"].map( mapping_dict )

In [50]:
print(laptops["os"].value_counts()) # see the change

Windows      1125
No OS          66
Linux          62
Chrome OS      27
macOS          21
Android         2
Name: os, dtype: int64


One important thing to remember with Series.map() is that if a value from your series doesn't exist as a key in your dictionary, it will convert that value to NaN

### Dropping missing values

In [51]:
laptops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 17 columns):
manufacturer         1303 non-null object
model_name           1303 non-null object
category             1303 non-null object
screen_size          1303 non-null object
screen               1303 non-null object
cpu                  1303 non-null object
ram                  1303 non-null object
storage              1303 non-null object
gpu                  1303 non-null object
os                   1303 non-null object
os_version           1133 non-null object
weight               1303 non-null object
price_euros          1303 non-null object
cpu_manufacture      1303 non-null object
gpu_manufacturer     1303 non-null object
screen_resolution    1303 non-null object
cpu_speed_ghz        1303 non-null float64
dtypes: float64(1), object(16)
memory usage: 173.1+ KB


In [108]:
print(laptops.isnull().sum())

manufacturer            0
model_name              0
category                0
screen_size_inches      0
screen                  0
cpu                     0
ram_gb                  0
storage                 0
gpu                     0
os                      0
os_version            170
weight_kg               0
price_euros             0
cpu_manufacture         0
gpu_manufacturer        0
screen_resolution       0
cpu_speed_ghz           0
dtype: int64




It's a lot clearer that we have only one column with null values– os_version, which has 170 missing values of the ~1300 total, about 13%.

We have a few options for how we can handle missing values:

    Remove any rows that have missing values.
    Remove any columns that have missing values.
    Fill the missing values with some other value.
    Leave the missing values as is.


In [52]:
laptops_no_null_rows = laptops.dropna( axis = 0 )

laptops_no_null_cols = laptops.dropna( axis =1 )
laptops_no_null_rows.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1133 entries, 5 to 1302
Data columns (total 17 columns):
manufacturer         1133 non-null object
model_name           1133 non-null object
category             1133 non-null object
screen_size          1133 non-null object
screen               1133 non-null object
cpu                  1133 non-null object
ram                  1133 non-null object
storage              1133 non-null object
gpu                  1133 non-null object
os                   1133 non-null object
os_version           1133 non-null object
weight               1133 non-null object
price_euros          1133 non-null object
cpu_manufacture      1133 non-null object
gpu_manufacturer     1133 non-null object
screen_resolution    1133 non-null object
cpu_speed_ghz        1133 non-null float64
dtypes: float64(1), object(16)
memory usage: 159.3+ KB


### Exploring the values  of Null columns  If something can be done apart from dropping the values

In [110]:
print(laptops["os_version"].value_counts(dropna=False))

10      1072
NaN      170
7         45
10 S       8
X          8
Name: os_version, dtype: int64




We can see that the majority of values in the column are 10, with the missing values the next most commons, and then about 5% of values being one of three others.

We can also explore values of the other columns in the rows with null values. In this case, the os_version column is closely related to the os column, so we'll look at those values:


In [111]:
os_with_null_v = laptops.loc[laptops["os_version"].isnull(),"os"]
print(os_with_null_v.value_counts())

No OS        66
Linux        62
Chrome OS    27
macOS        13
Android       2
Name: os, dtype: int64




Immediately we can observe a few things:

    Most of the missing values are actually when the laptop doesn't include any OS. This is an important distinction, because it's not so much that we don't know what the value is, as that there can't be a value.
    13 of the laptops that come with macOS do not specify the version. Leaning on our knowledge of MacOS, we might know that the full name of macOS used to be Mac OS X, and so we might to fill these values to be more consistent.

In both of these cases, we can fill the missing values to make our data more correct. For the rest of the values, it's probably best to leave them as missing so we don't remove important values.

First, let's explore those mac columns a bit more to make sure our intuition was correct:


In [112]:
mac_os_versions = laptops.loc[laptops["os"] == "macOS", "os_version"]
print(mac_os_versions.value_counts(dropna=False))

NaN    13
X       8
Name: os_version, dtype: int64


We were correct, all 8 rows that have the value X are macOS versions. We'll fill in all of the NaN values with X. We can use assignment with a boolean comparison to perform this replacement:


For our other case, let's insert a Version Unknown value into the os_version column for any laptop with a No OS value in the os column:

instructions

    Use a boolean array to identify rows that have the value No OS for the os column, and then use assignment to assign the value Version Unknown to the os_version column for those rows.

    Use the syntax below to create value_counts_after variable:

    value_counts_after = laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()

    After you have run your code, use the variable inspector to look at the difference between value_counts_before and value_counts_after.


In [113]:
value_counts_before = laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()
laptops.loc[laptops["os"] == "macOS", "os_version"] = "X"


# 
laptops.loc[laptops["os"] == 'No OS', "os_version"] = "Version Unknown" 

value_counts_after = laptops.loc[laptops["os_version"].isnull(),"os"].value_counts()

### Cleaning Storage

In [114]:
print(laptops.loc[76:81, "storage"])

76                 2TB HDD
77    128GB SSD +  1TB HDD
78                 1TB HDD
79    128GB SSD +  1TB HDD
80               256GB SSD
81               512GB SSD
Name: storage, dtype: object


In [123]:
all_storage = laptops["storage"].str.rsplit(n=2, expand=True).iloc[:,0]
type(all_storage) # should be series

pandas.core.series.Series

In [124]:
all_storage.unique()

array(['128GB', '256GB', '512GB', '500GB', '1TB', '32GB', '128GB SSD +',
       '256GB SSD +', '64GB', '2TB', '512GB SSD +', '16GB',
       '64GB Flash Storage +', '1GB', '1TB HDD +', '1TB SSD +', '240GB',
       '8GB', '508GB'], dtype=object)

In [53]:
# replace 'TB' with 000 and rm 'GB'
laptops["storage"] = laptops["storage"].str.replace('GB','').str.replace('TB','000')

# split out into two columns for storage
laptops[["storage_1", "storage_2"]] = laptops["storage"].str.split("+", expand=True)

for s in ["storage_1", "storage_2"]:
    s_capacity = s + "_capacity_gb"
    s_type = s + "_type"
    # create new cols for capacity and type
    laptops[[s_capacity, s_type]] = laptops[s].str.split(n=1,expand=True)
    # make capacity numeric (can't be int because of missing values)
    laptops[s_capacity] = laptops[s_capacity].astype(float)
    # strip extra white space
    laptops[s_type] = laptops[s_type].str.strip()

# remove unneeded columns
laptops.drop(["storage", "storage_1", "storage_2"], axis=1, inplace=True)

In [54]:
laptops_dtypes = laptops.dtypes
cols = ['manufacturer', 'model_name', 'category', 'screen_size_inches',
        'screen', 'cpu', 'cpu_manufacturer',  'cpu_speed', 'ram_gb',
        'storage_1_type', 'storage_1_capacity_gb', 'storage_2_type',
        'storage_2_capacity_gb', 'gpu', 'gpu_manufacturer', 'os',
        'os_version', 'weight_kg', 'price_euros']
        
        
laptops = laptops[cols]

laptops.to_csv('laptops_cleaned.csv',index=False)
laptops_cleaned = pd.read_csv('laptops_cleaned.csv')
laptops_cleaned_dtypes = laptops_cleaned.dtypes        

KeyError: "['screen_size_inches' 'cpu_manufacturer' 'cpu_speed' 'ram_gb' 'weight_kg'] not in index"